src/main/java/net/glowstone/block/data/state/StateGenerator.java

```diff
@@ -0,0 +1,102 @@
+package net.glowstone.block.data.state;
+
+import net.glowstone.block.data.state.generator.BooleanStateGenerator;
+import net.glowstone.block.data.state.generator.EnumStateGenerator;
+import net.glowstone.block.data.state.generator.IntegerStateGenerator;
+import org.bukkit.Axis;
+import org.bukkit.Instrument;
+import org.bukkit.block.BlockFace;
+import org.bukkit.block.data.Bisected;
+import org.bukkit.block.data.Rail;
+import org.bukkit.block.data.type.*;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public interface StateGenerator<T> {
```

Comment:
What is this interface being used for? Is it attempting to generate the index in the block report's property value arrays within the state files?

Reply:
The StateGenerator class is for the common class between all generators. So the Generators are used to generate all common values for the states, such as network codes, the ID of the StateValue. 

So for example, the NETHER_PORTAL block data has a state of AXIS that can be either X or Z. The StateValue will contain the current value of the AXIS, as in it will either hold the enum of Axis.X or Axis.Z, while the generator will hold all the possible values for the AXIS nether portal, as well hold the ID of AXIS.

minecraft:nether_portal{axis=x}
[material id]{[generator]=[value]}

Reply:
Our plan is currently to utilize the blocks report generated by the vanilla server JAR at runtime. Due to this, there will likely be a separate manager handling the downloading of the vanilla server and generation of the report (if not already present), and desalinizing of the JSON model. I've got a prototype manager written already that reads and deserializes this report. How would this model interface with such a mechanism for determining the network IDs?

Reply:
From what I can remember about those JSON files. It contains the state value as its network id as well as the material. Right? 

If so, then you would create the default BlockData interface for your material, get the generators for the BlockData and then create the StateValue of the network ID from the generator and replace the default value in the BlockData's StateValue for that ID to the newly created StateValue. 

I havent fully got the network system in place yet, but you can see some of it such as 

StateGenerator.deserialize(int networkId)

Reply:
I documented the structure of that file here using JSON Schema: https://github.com/GlowstoneMC/1.13-board/issues/28#issuecomment-575973492

Basically, each combination of property values is mapped to its own block ID. Within this file, all values are strings, so the most efficient way to find the block IDs would be to look them up based on the block type and a Map<String, String> of properties.

Reply:
Wouldn't it be more effectient to have a map value of Map<String, BlockData>? Im assuming when you mean Map<String, String> you mean the unique id that these property values and the mapped string is the typical string id, as in "minecraft:END_ROD{face=NORTH}"?

Reply:
No, in your example I literally mean the structure is (not real data from the file):

 ```
{
    "minecraft:end_rod": {
        "properties": {
            "face": ["NORTH", "SOUTH", "EAST", "WEST"]
        },
        "states": [
            {
                "properties": {
                    "face": "NORTH"
                 },
                 "id": 123,
                 "default": true
            },
            {
                "properties": {
                    "face": "SOUTH"
                 },
                 "id": 124
            },
            {
                "properties": {
                    "face": "EAST"
                 },
                 "id": 125
            },
            {
                "properties": {
                    "face": "WEST"
                 },
                 "id": 126
            },
        ]
    }
}
```

In my explanation above, the Map<String, String> corresponded to obj["minecraft:end_rod"].states[0].properties, for example. While, yes, I could parse this out all the way into BlockData, there would be quite a lot of operations being done just to figure out every state of every block (again, all combinations for each block are represented within this file). Instead, my code just parses it into the bare minimum needed for you to be able to match a block's state to its network ID

Reply:
Oh, and by the way, I couldn't even use the NBT functions to create the BlockData for each state in this file, since the file represents all values using JSON, not NBT. We'd need to maintain a separate function, just for unfolding this.

Reply:
We have a Mojangson class that can help.

Reply:
The problem isn't with parsing this stuff, the problem is with translating the model from the raw form above into the BlockData that Bukkit expects. We'd be having to maintain a separate method full of switch cases for doing this mapping.

Reply:
Right. Ok, I can handle that

Reply:
Is this what you had in mind? @smartboyathome 

3efe441c7d4cde4ff57368231508b600dd8d46af

Reply:
Not quite what I had in mind. When I get a chance, I'll send you over the interfaces I was coding against. With this method, I am seriously concerned we'll end up past Java's 64KB block limit eventually.